File: src/Originations/LineOfCredit.java
package Originations;

import java.text.*;
import java.util.*;

public class LineOfCredit {

	private final static double MONTHLY_FEE = 6.25;
	private final static double INTEREST_RATE = 0.35;
	private final static double DRAW_FEE_PCT = 0.03;
	private final static double DRAW_FEE_MIN = 10.00;
	private final static int DUE_AFTER = 14;
	
	private int id;
	private String appNumber;
	private int userId;
	private String email;
	private double creditLine;
	private Date openDate;
	private Date firstDueDate;
	
	private Ledger ledger = new Ledger();
	private Map<Date,Double> payments = new HashMap<Date,Double>();

	private Calendar interestCalendar;
	private Calendar billingCalendar;
	private boolean hadDraw = false;
	
	public LineOfCredit(Date date, int id, String appNumber, int userId, String email, double creditLine, Date firstDueDate) {
		this.id = id;
		this.appNumber = appNumber;
		this.userId = userId;
		this.email = email;
		this.openDate = date;
		this.creditLine = creditLine;
		this.firstDueDate = firstDueDate;
		interestCalendar = Calendar.getInstance();
		interestCalendar.setTime(date);
		billingCalendar = Calendar.getInstance();
		billingCalendar.setTime(date);
	}
	
	/* getters */
	public int getId() { return id; }
	public int getUserId() { return userId; }
	public String getAppNumber() { return appNumber; }
	public String getEmail() { return email; }
	public double getCreditLine() { return creditLine; }
	public Date getOpenDate() {return openDate; }


	/* fees */
	public double getMonthlyFee() { return MONTHLY_FEE; }
	
	public double getDrawFee(double amount) {
		return Math.max(DRAW_FEE_MIN, amount * DRAW_FEE_PCT);
	}
	
	/* ledger amounts */
	public double getOutstanding(Date date) { return ledger.getPrincipal(date); }

	public double getFees(Date date) { return ledger.getFees(date); }	
	
	public double getAccruedInterest(Date date) {
		moveForwardTo(date);
		return ledger.getInterest(date);
	}	
	
	public double getPayoff(Date date) {
		moveForwardTo(date);
		return ledger.getBalance(date);
	}	
	
	/* draws */
	public void draw(Date date, double amount) {
		moveForwardTo(date);
		if (!hadDraw) {
			ledger.addEntry("Fee", billingCalendar.getTime(), getMonthlyFee(), true);
			hadDraw = true;
		}
		ledger.addEntry("Principal", date, amount, true);
		ledger.addEntry("Fee", date, getDrawFee(amount), true);
	}

	/* interest accrual */
	public double getDailyInterest(Date date) {
		return INTEREST_RATE/365*ledger.getPrincipal(date);
	}	
	
	public void accrueInterest(Date date) {
		while (interestCalendar.getTime().before(date)) {
			double interest = getDailyInterest(interestCalendar.getTime());
			interestCalendar.add(Calendar.DATE, 1);
			ledger.addEntry("Interest", interestCalendar.getTime(), interest, true);
		}
	}

	/* payments */
	public void pay(Date date, double amount) {
		moveForwardTo(date);
		payments.put(date, Double.valueOf(amount));
		double fees = Math.min(ledger.getFees(date), amount);
		ledger.addEntry("Fee", date, fees, false);
		amount -= fees;
		double interest = Math.min(ledger.getInterest(date), amount);
		ledger.addEntry("Interest", date, interest, false);
		amount -= interest;
		ledger.addEntry("Principal", date, amount, false);
	}	
	
	/* billing cycles */
	public Date skipWeekEnd(Date date) {
		Calendar cal = Calendar.getInstance();
		cal.setTime(date);
		switch (cal.get(Calendar.DAY_OF_WEEK)) {
			case Calendar.SATURDAY: cal.add(Calendar.DATE, 2); break;
			case Calendar.SUNDAY: cal.add(Calendar.DATE, 1); break;
		}
		return cal.getTime();
	}
	
	public Date getNextDueDate(Date date) {
		Calendar cal = Calendar.getInstance();
		cal.setTime(firstDueDate);
		while (!skipWeekEnd(cal.getTime()).after(date)) {
			cal.add(Calendar.MONTH, 1);
		}
		return skipWeekEnd(cal.getTime());
	}

	public Date getNextDueDate() {
		return skipWeekEnd(getNextDueDate(billingCalendar.getTime()));
	}
	
	public Date getPreviousDueDate(Date date) {
		Calendar cal = Calendar.getInstance();
		cal.setTime(firstDueDate);
		Date previousDueDate = cal.getTime();
		while (!skipWeekEnd(cal.getTime()).after(date)) {
			previousDueDate = cal.getTime();
			cal.add(Calendar.MONTH, 1);
		}
		return skipWeekEnd(previousDueDate);
	}

	public Date getPreviousDueDate() {
		return skipWeekEnd(getNextDueDate(billingCalendar.getTime()));
	}
	
	public Date getNextStatementDate(Date date) {
		Calendar cal = Calendar.getInstance();
		cal.setTime(getNextDueDate(date));
		cal.add(Calendar.DATE, -DUE_AFTER);
		if (!cal.getTime().after(date)) {
			cal.setTime(getNextDueDate(getNextDueDate(date)));
			cal.add(Calendar.DATE, -DUE_AFTER);
		}
		return cal.getTime();
	}

	public Date getNextStatementDate() {
		return getNextStatementDate(billingCalendar.getTime());
	}
	
	public Date getPreviousStatementDate(Date date) {
		Calendar cal = Calendar.getInstance();
		cal.setTime(getNextDueDate(date));
		cal.add(Calendar.DATE, -DUE_AFTER);
		if (cal.getTime().after(date)) {
			cal.setTime(getPreviousDueDate(date));
			cal.add(Calendar.DATE, -DUE_AFTER);
		}
		return cal.getTime();
	}

	public Date getPreviousStatementDate() {
		return getPreviousStatementDate(billingCalendar.getTime());
	}
	
	public void addBillingCycles(Date date) {
		while (!getNextStatementDate(billingCalendar.getTime()).after(date)) {
			billingCalendar.setTime(getNextStatementDate(billingCalendar.getTime()));
			if (hadDraw) ledger.addEntry("Fee", billingCalendar.getTime(), getMonthlyFee(), true);
		}
	}
	
	public void moveForwardTo(Date date) {
		accrueInterest(date);
		addBillingCycles(date);
	}
	
	/* minimum payment */
	public double getLastDrawOutstanding(Date date) {
		return ledger.getLastDrawPrincipal(date);
	}

	public double getDenominator(double line) {
		double denominator;
		if (this.creditLine <= 350) denominator = 5;
		else if (this.creditLine <= 500) denominator = 6;
		else if (this.creditLine <= 750) denominator = 8;
		else denominator = 10;
		return denominator;
	}
	
	public double getMinPayment(Date date) {
		return getLastDrawOutstanding(date) / getDenominator(this.creditLine);
	}
	
	public double getPayments(Date start, Date end) {
		double paid = 0;
		for (Date day : payments.keySet()) {
			if (!day.before(start) && !day.after(end)) paid += payments.get(day).doubleValue();
		}
		return paid;
	}
	
	public double getPaymentDue(Date date) {
		Calendar cal = Calendar.getInstance();
		Date lastStatementDate = getPreviousStatementDate(date);
		cal.setTime(lastStatementDate);
		cal.add(Calendar.DATE, DUE_AFTER);
		double paid = getPayments(lastStatementDate, cal.getTime()); 
		double minPayment = Math.max(0, getMinPayment(date) - paid);
		return Math.min(minPayment, getOutstanding(date));
	}
	
	/* diagnostic info */
	public String toString(Date date) {
		NumberFormat twoFixed = new DecimalFormat("0.00");
		SimpleDateFormat mdyyyy = new SimpleDateFormat("M/d/yyyy");
		return
			Integer.toString(id) +
			"\t" + appNumber +
			"\t" + Integer.toString(userId) +
			"\t" + String.format("%-30s", email) +
			"\t" + twoFixed.format(creditLine) +
			"\t" + twoFixed.format(getOutstanding(date)) +
			"\t" + twoFixed.format(getFees(date)) +
			"\t" + twoFixed.format(getAccruedInterest(date)) +
			"\t" + twoFixed.format(getPayoff(date)) +
			"\t" + mdyyyy.format(getPreviousDueDate(date)) +
			"\t" + mdyyyy.format(getNextDueDate(date)) +
			"\t" + twoFixed.format(getPaymentDue(date));
	}

	public String toString() {
		return toString(Calendar.getInstance().getTime());
	}
	
	public String getHeaderString() {
		return
			"ID" +
			"\t" + "AppNumber" +
			"\t" + "UserID" +
			"\t" + String.format("%-30s", "Email") +
			"\t" + "Line" +
			"\t" + "Out" +
			"\t" + "Fees" +
			"\t" + "Int" +
			"\t" + "Payoff" +
			"\t" + "PrevDate" +
			"\t" + "NextDate" +
			"\t" + "Due";
	}
	
	public void printLedger() {
		ledger.print();
	}
	
}

File: lib/LOC/LineOfCredit.rb

module LOC

	class LineOfCredit 
	
		MONTHLY_FEE = 6.25
		INTEREST_RATE = 0.35
		DRAW_FEE_PCT = 0.03
		DRAW_FEE_MIN = 10.00
		DUE_AFTER = 14

		COL_WIDTHS = [11, 5, 11, 30, 11, 7, 7, 7, 7, 7, 11, 11, 7, 7, 11, 7]
		COL_JUST   = [1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0]
		COL_NAMES  = ["AsOfDate", "ID", "AppNumber", "Email", "Opened", "Line", "Pr.", "Fees", "Int.", "Balance", "PrevDue", "NextDue", "MinPay", "AmtDue", "LastPaid", "PaidAmt"]
		HEADER = COL_NAMES.each_with_index.map { |s, i|
			(COL_JUST[i] == 1) ? s.to_s[0..COL_WIDTHS[i]-1].ljust(COL_WIDTHS[i]) : s.to_s[0..COL_WIDTHS[i]-1].rjust(COL_WIDTHS[i])
		}.join(" ")
		
		attr_accessor :id, :app_number, :user_id, :email, :credit_line, :open_date
		attr_accessor :ledger, :payments
		attr_accessor :last_billing_date, :last_interest_date
		
		def initialize(open_date, id, app_number, user_id, email, credit_line, first_due_date) 
			@ledger = Ledger.new
			@payments = {}
			@id = id
			@app_number = app_number
			@user_id = user_id
			@email = email
			@open_date = open_date
			@credit_line = credit_line
			@first_due_date = first_due_date
			@last_interest_date = open_date
			@last_billing_date = open_date
			@had_draw = false
		end
		
		# fees
		def monthly_fee
			MONTHLY_FEE
		end
		
		def draw_fee(amount) 
			[DRAW_FEE_MIN, amount * DRAW_FEE_PCT].max
		end
		

		# ledger amounts
		def principal(date)
			@ledger.balance(LedgerEntry::PRINCIPAL, date)
		end
	
		def fees(date)
			@ledger.balance(LedgerEntry::FEE, date)
		end
	
		def accrued_interest(date)
			move_forward_to(date)
			@ledger.balance(LedgerEntry::INTEREST, date)
		end
	
		def total_balance(date)
			move_forward_to(date)
			@ledger.balance(LedgerEntry::TOTAL, date)
		end
	
		
		# draws
		def draw(date, amount) 
			if (!@had_draw && amount > 0) 
				@ledger.add(LedgerEntry::FEE, "Monthly fee", @last_billing_date, monthly_fee, LedgerEntry::CREDIT)
				@had_draw = true
			end
			move_forward_to(date)
			@ledger.add(LedgerEntry::PRINCIPAL, "Transfer", date, amount, LedgerEntry::CREDIT)
			@ledger.add(LedgerEntry::FEE, "Tranfer fee", date, draw_fee(amount), LedgerEntry::CREDIT)
		end

	
		# interest accrual
		def daily_interest(date) 
			return INTEREST_RATE/365*principal(date)
		end	
		
		def accrue_interest(date) 
			while (@last_interest_date < date) 
				i = daily_interest(@last_interest_date)
				@last_interest_date += 1
				@ledger.add(LedgerEntry::INTEREST, "Accrual  ", @last_interest_date, i, LedgerEntry::CREDIT)
			end
		end

	
		# payments 
		def pay(date, amount)
			move_forward_to(date)
			@payments[date] = amount
			f = [fees(date), amount].min
			@ledger.add(LedgerEntry::FEE, "Payment  ", date, f, LedgerEntry::DEBIT)
			amount -= f
			i = [accrued_interest(date), amount].min
			@ledger.add(LedgerEntry::INTEREST, "Payment  ", date, i, LedgerEntry::DEBIT)
			amount -= i
			@ledger.add(LedgerEntry::PRINCIPAL, "Payment  ", date, amount, LedgerEntry::DEBIT)
		end	

		
		# billing cycles 
		def skip_weekend(date) 
			(date.wday == 6) ? date + 2 : (date.wday == 0) ? date + 1 : date
		end
		
		def next_due_date(date = @last_billing_date) 
			d = @first_due_date
			d = d >> 1 while skip_weekend(d) <= date 
			return skip_weekend(d)
		end

		def prev_due_date(date = @last_billing_date) 
			d = @first_due_date
			d = d >> 1 while skip_weekend(d) <= date 
#			return skip_weekend(d << 1)
			return (@first_due_date <= (d << 1)) ? skip_weekend(d << 1) : nil
		end
	
		
		def next_statement_date(date = @last_billing_date) 
			d = next_due_date(date) - DUE_AFTER
			d = next_due_date(next_due_date(date)) - DUE_AFTER if (d <= date)
			return d
		end
		
		def prev_statement_date(date = @last_billing_date) 
			d = next_due_date(date) - DUE_AFTER
			if (d > date) then
				d = prev_due_date(date).nil? ? nil : prev_due_date(date) - DUE_AFTER
			end
			return d
		end
	
		def generate_statements(date) 
			while next_statement_date(@last_billing_date) <= date 
				@last_billing_date = next_statement_date(@last_billing_date)
				ledger.add(LedgerEntry::FEE, "Monthly fee", @last_billing_date + 1, monthly_fee, LedgerEntry::CREDIT) if (@had_draw)
			end
		end

		def move_forward_to(date) 
			accrue_interest(date)
			generate_statements(date)
		end
		
		# minimum payment
		def denominator
			case 
				when @credit_line <= 350 then 5
				when @credit_line <= 500 then 6
				when @credit_line <= 750 then 8
				else 10
			end
		end
		
		def min_payment(date = @last_billing_date)
			@ledger.outstanding_last_draw(date||@last_billing_date) / denominator
		end

		
		def paid(start_date, end_date) 
			p = 0
			@payments.each do |day, amount|
				p += amount if day >= start_date && day <= end_date
			end
			return p
		end
		
		def payment_due(date = @last_interest_date) 
			p = [0, min_payment(prev_statement_date(date)) - paid(prev_statement_date(date), date)].max
			return [p, total_balance(date)].min
		end
	
		
		# display
		def pad(s, i)
			if COL_JUST[i] == 1 then
				s.to_s[0..COL_WIDTHS[i]-1].ljust(COL_WIDTHS[i])
			else
				s.to_s[0..COL_WIDTHS[i]-1].rjust(COL_WIDTHS[i])
			end
		end

		def to_s(date = @last_interest_date)
			[date,
			 @id,
			 @app_number,
			 @email,
			 @open_date,
			 '%.2f' % @credit_line,
			 '%.2f' % principal(date),
			 '%.2f' % fees(date),
			 '%.2f' % accrued_interest(date),
			 '%.2f' % total_balance(date),
			 prev_due_date(date),
			 next_due_date(date),
			 '%.2f' % min_payment(prev_statement_date(date)),
			 '%.2f' % payment_due(date),
			 @payments.keys.last,
			 @payments.values.last.nil? ? "" : '%.2f' % @payments.values.last
			].each_with_index.map { |s, i|
				(COL_JUST[i] == 1) ? s.to_s[0..COL_WIDTHS[i]-1].ljust(COL_WIDTHS[i]) : s.to_s[0..COL_WIDTHS[i]-1].rjust(COL_WIDTHS[i])
			}.join(" ")
		end

		def to_csv(date = @last_interest_date)
			[date,
			 @id,
			 @app_number,
			 @email,
			 @open_date,
			 '%.2f' % @credit_line,
			 '%.2f' % principal(date),
			 '%.2f' % fees(date),
			 '%.2f' % accrued_interest(date),
			 '%.2f' % total_balance(date),
			 prev_due_date(date),
			 next_due_date(date),
			 '%.2f' % min_payment(prev_statement_date(date)),
			 '%.2f' % payment_due(date),
			 @payments.keys.last,
			 @payments.values.last.nil? ? "" : '%.2f' % @payments.values.last
			].join(",")
		end


	end

end
